File: 2025/problems/sovietunionfind/submissions/accepted/atli.py
import sys


class SovietUF:
    def __init__(self, n):
        self.p = [-1 for _ in range(n)]
        self.nxt = [-1 for _ in range(n)]
        self.fst = [i for i in range(n)]
        self.lst = [i for i in range(n)]

    def find(self, x):
        while self.p[x] >= 0:
            x, self.p[x] = self.p[x], self.p[self.p[x]]
        return x

    def unite(self, x, y):
        xp = self.find(x)
        yp = self.find(y)
        if xp == yp:
            return
        swp = False
        if self.p[xp] > self.p[yp]:
            xp, yp = yp, xp
            swp = True
        self.p[xp] += self.p[yp]
        self.p[yp] = xp
        if swp:
            self.nxt[self.lst[yp]] = self.fst[xp]
            self.fst[xp] = self.fst[yp]
        else:
            self.nxt[self.lst[xp]] = self.fst[yp]
            self.lst[xp] = self.lst[yp]

    def balkanise(self, x):
        xp = self.find(x)
        todo = [self.fst[xp]]
        while self.nxt[todo[-1]] != -1:
            todo.append(self.nxt[todo[-1]])
        for y in todo:
            self.p[y] = -1
            self.fst[y] = y
            self.lst[y] = y
            self.nxt[y] = -1


n, q = map(int, sys.stdin.readline().strip().split())
suf = SovietUF(n)
for _ in range(q):
    inp = input().strip().split()
    if inp[0] == "a":
        x, y = map(int, inp[1:])
        suf.unite(x - 1, y - 1)
    elif inp[0] == "b":
        suf.balkanise(int(inp[1]) - 1)
    elif inp[0] == "c":
        p = suf.find(int(inp[1]) - 1)
        sys.stdout.write(str(suf.fst[p] + 1) + "\n")

File: 2025/problems/sovietunionfind/submissions/accepted/atli.cpp
#include<cassert>
#include<iostream>
#include<list>
#include<vector>
using namespace std;
typedef vector<int> vi;

struct sovietuf {
    vi p; vector<list<int>> mem;
    sovietuf(int n) : p(n, -1) {
        mem = vector<list<int>>(n);
        for(int i = 0; i < n; ++i)
            mem[i].push_back(i);
    }
    int find(int x) {
        return p[x] < 0 ? x : p[x] = find(p[x]);
    }
    void unite(int x, int y) {
        int a = find(x), b = find(y);
        if(a == b) return;
        bool swp = false;
        if(p[a] > p[b]) swap(a, b), swp = true;
        p[a] += p[b];
        p[b] = a;
        if(swp) {
            mem[a].splice(mem[a].begin(), mem[b]);
        } else {
            mem[a].splice(mem[a].end(), mem[b]);
        }
    }
    void balkanise(int x) {
        int a = find(x);
        vi todo(mem[a].begin(), mem[a].end());
        for(int y : todo) {
            p[y] = -1;
            mem[y] = list<int>();
            mem[y].push_back(y);
        }
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int n, q; cin >> n >> q;
    sovietuf suf(n);
    while(q--) {
        char op;
        cin >> ws >> op;
        if(op == 'a') {
            int x, y;
            cin >> x >> y;
            x--; y--;
            suf.unite(x, y);
        } else if(op == 'b') {
            int x; cin >> x; x--;
            suf.balkanise(x);
        } else if(op == 'c') {
            int x; cin >> x; x--;
            int p = suf.find(x);
            cout << *suf.mem[p].begin() + 1 << '\n';
        } else assert(false);
    }
}
